File: app/src/main/java/com/kakeibo/data/SubscriptionContent.java
package com.kakeibo.data;

import android.os.Parcel;
import android.os.Parcelable;

import androidx.annotation.Nullable;

public class SubscriptionContent implements Parcelable {
    @Nullable
    public final String title;
    @Nullable
    public final String subtitle;
    @Nullable
    public final String description;

    @Override
    public int describeContents() {
        return 0;
    }

    @Override
    public void writeToParcel(Parcel dest, int flags) {
        dest.writeString(title);
        dest.writeString(subtitle);
        dest.writeString(description);
    }

    public static final Creator<SubscriptionContent> CREATOR
            = new Creator<SubscriptionContent>() {
        public SubscriptionContent createFromParcel(Parcel in) {
            return new SubscriptionContent(in);
        }

        public SubscriptionContent[] newArray(int size) {
            return new SubscriptionContent[size];
        }
    };

    private SubscriptionContent(Parcel in) {
        title = in.readString();
        subtitle = in.readString();
        description = in.readString();
    }

    private SubscriptionContent(@Nullable String title,
                                @Nullable String subtitle,
                                @Nullable String desc) {
        this.title = title;
        this.subtitle = subtitle;
        this.description = desc;
    }

    class Builder {
        private String title = null;
        private String subTitle = null;
        private String description = null;

        Builder setTitle(String title) {
            this.title = title;
            return this;
        }

        Builder setSubtitle(String subtitle) {
            this.subTitle = subtitle;
            return this;
        }

        Builder setDescription(String description) {
            this.description = description;
            return this;
        }

        public SubscriptionContent build() {
            return new SubscriptionContent(title, subTitle, description);
        }
    }
}

File: app/src/main/java/com/kakeibo/feature_subscriptions/SubscriptionContent.kt
package com.kakeibo.data

import android.os.Parcel
import android.os.Parcelable

class SubscriptionContent : Parcelable {
    val title: String?
    val subtitle: String?
    val description: String?

    override fun describeContents(): Int {
        return 0
    }

    override fun writeToParcel(dest: Parcel, flags: Int) {
        dest.writeString(title)
        dest.writeString(subtitle)
        dest.writeString(description)
    }

    private constructor(input: Parcel) {
        title = input.readString()
        subtitle = input.readString()
        description = input.readString()
    }

    private constructor(
            title: String?,
            subtitle: String?,
            desc: String?
    ) {
        this.title = title
        this.subtitle = subtitle
        description = desc
    }

    override fun toString() =
            "SubscriptionContent{title='$title', subtitle='$subtitle', description='$description'}"

    internal inner class Builder {
        private var title: String? = null
        private var subTitle: String? = null
        private var description: String? = null

        fun setTitle(title: String): Builder {
            this.title = title
            return this
        }

        fun setSubtitle(subtitle: String): Builder {
            this.subTitle = subtitle
            return this
        }

        fun setDescription(description: String?): Builder {
            this.description = description
            return this
        }

        fun build(): SubscriptionContent {
            return SubscriptionContent(title, subTitle, description)
        }
    }

    companion object CREATOR: Parcelable.Creator<SubscriptionContent> {
        override fun createFromParcel(parcel: Parcel): SubscriptionContent {
            return SubscriptionContent(parcel)
        }

        override fun newArray(size: Int): Array<SubscriptionContent?> {
            return arrayOfNulls(size)
        }
    }
}